ImportXml.py

```diff
@@ -0,0 +1,182 @@
+# import pandas as pd
+# from collections import defaultdict
+
+
+def read_questionnaire_structure(filepath: str, writefiles: bool):
+    """
+    Reads questionnaire structure XML file
+
+    Args:
+        filepath: A string with a path to a XML file with questionnaire structure
+        writefiles: if True will dumps data to files as .json which can be opened \
+                    by json.load(file) or pandas.read_json(file).
+    Returns:
+        A dict containing the questionnaire structure:
+        {
+            "sections": {
+               {
+                   "id":13901,
+                   "title": "Demographics",
+                   "info": "In this section, we will ask general questions about yourself \
+                       and your doctoral project."
+               },
+               ....
+            },
+            "questions": {
+                "name": "A1",
+                "question": "What is your year of birth?",
+                "description": "Please use the following format: 'xxxx' , e.g. 1989",
+                "type": "category",
+                "categories": {"I don't want to answer this question": 'A2',
+                                ...
+                                'Year of birth: ': 'A4other'},
+                "subquestions": {
+                    'E7a_SQ001': 'My supervisor is well informed about my field of research.',
+                    'E7a_SQ002': 'My supervisor is available when I need advice.',
+                    'E7a_SQ003': 'My supervisor is open to and respects my research ideas.',
+                    'E7a_SQ004': 'My supervisor gives constructive feedback.',
+                    ...},
+                "section_id": 13901
+            }
+        }
+
+    Raises:
+        ValueError: Invalid XML file
+        ValueError: Provided file has unexpected structure
+    """
+
+    def uniCodeRe(s) -> str:
+        # '\u00a0','\u20ac', '\u00b2', '\u00a7'
+        matched = ["\n", "\t", "\xa0"]
+        # if you want to change the unicode
+        # s_new = s.replace(u'\u2018', '"').replace(u'\u2019', '\'').replace(u'\u00f6', r'ö').replace(u'\u00fc', r'ü')
+        # s_new = s
+        res_s = re.sub("|".join(matched), "", s)
+        return res_s
+
+    def textParse(bsString) -> str:
+        text = ""
+        sps = BeautifulSoup(bsString, "html.parser").find_all("p")
+        # sbs = BeautifulSoup(bsString, 'html.parser').find_all('b')
+        for sp in sps:
+            if bool(sp.string):
+                text += uniCodeRe(sp.string)
+        return text
+
+    def findCategories(resp) -> dict:
+        cates = dict()
+        cats = resp.find_all("category")
+        for cat in cats:
+            cat_lab = cat.label.string
+            cat_val = cat.value.string
+            # A#other case
+            if bool(cat.contingentQuestion):
+                cat_val = cat.contingentQuestion.attrs["varName"]
+                cat_lab = uniCodeRe(cat.contingentQuestion.find("text").get_text())
```

Comment:
this part is used for parse the tag "contigentQuestion" which is I comment before, about the "A1other", "A2other" ... these multiple choice categories. In order to parse these part out. But these "other" categories are not same as other categories of under one question. Normally the categories under one question used labe and value to represent, but these "other" category under the same question is labeled by this "contigentQuestion" tag, and the label name and attr are also not same. Moreover, not all of the question has these "other" categories. so I need this if condition to choose if it is a "other" kind, find it correspond label and value.

Reply:
Sorry, I don't get it:

![image](https://user-images.githubusercontent.com/9852534/133573776-70e61783-bdd1-4d89-b983-463449f6c4cf.png)

Here we have "A2other". If you take `label` and `value` it would be `label='Other'` and `value='-oth-'`. But in the current version of code it will be `label='Other'` and `value='A2other'`. Do I understand it right? Why do you want to put "A2other" as category value?

Reply:
ye, it's correct, sorry I missed there already exists a <label> tag under <category> tag, so I use the <text> tag's content "other". But I think this "if contingentQuesiton" condition is still needed, as some question don't have this "other" category.